onnxruntime/core/session/onnxruntime_c_api_ep.cc

```diff
@@ -0,0 +1,733 @@
+        if (!graph_build.GetInitializedTensor(input->Name(), subgraph_initializer)) {
+          graph_build.AddInitializedTensor(*(initializer));
+        }
+      }
+    }
+    for (auto input : node->ImplicitInputDefs()) {
+      const ONNX_NAMESPACE::TensorProto* initializer = nullptr;
+      if (graph_viewer->GetInitializedTensor(input->Name(), initializer)) {
+        const ONNX_NAMESPACE::TensorProto* subgraph_initializer = nullptr;
+        if (!graph_build.GetInitializedTensor(input->Name(), subgraph_initializer)) {
+          graph_build.AddInitializedTensor(*(initializer));
+        }
+      }
+    }
+    for (auto output : node->OutputDefs()) {
+      auto& n_output = graph_build.GetOrCreateNodeArg(output->Name(), output->TypeAsProto());
+      outputs.push_back(&n_output);
+      const auto name = output->Name();
+      if (graph_output_names.find(name) != graph_output_names.end()) {
+        subgraph_output_names.push_back(name);
+      }
+    }
+
+    std::unordered_set<std::string> control_flow_op_set = {"If", "Loop", "Scan"};
+    if (control_flow_op_set.find(node->OpType()) != control_flow_op_set.end()) {
+      has_control_flow_op = true;
+    }
+
+    // If the node has subgraph, it's possible that the ORT graph of that subgraph and the GraphProto in the node attributes are not in sync because of graph optimization.
+    // Therefore, we need to force GraphProto attributes to be updated in order to get the valid GraphProto.
+    if (node->GetAttributes().size() > 0) {
+      auto node_proto = std::make_unique<ONNX_NAMESPACE::NodeProto>();
+      // we need to update any GraphProto attributes for subgraphs so that any changes made by things
+      // such as the optimizers are captured. otherwise we can end up saving an invalid graph.
+      node->ToProto(*node_proto, /* update_subgraphs */ true);
+      const int num_attributes = node_proto->attribute_size();
+      NodeAttributes node_attributes;
+      node_attributes.reserve(num_attributes);
+
+      for (int i = 0; i < num_attributes; ++i) {
+        auto& attr = node_proto->attribute(i);
+        node_attributes.emplace(attr.name(), attr);
+      }
+
+      // The GraphProto attributes are the updated ones.
+      graph_build.AddNode(node->Name(), node->OpType(), node->Description(), inputs, outputs, &node_attributes, node->Domain());
+    } else {
+      // The GraphProto attributes are the original ones.
+      graph_build.AddNode(node->Name(), node->OpType(), node->Description(), inputs, outputs, &node->GetAttributes(), node->Domain());
+    }
+  }
+
+  // TODO:yang
+  // Only if the newly built graph has control flow op as well as it has parent node,
+  // it needs to handle outer scope values before calling graph.Resolve().
+  // TODO(leca): Is local variable enough? Do we need to make it EP class variable?
+  std::unordered_map<std::string, std::unique_ptr<SubGraphContext2>> subgraph_context_map;
```

Comment:
I think making `subgraph_context_map` as a local variable should be enough.
The implication of doing so is that the same context of a specific subgraph might be built and release multiple times across multiple `OrtGraphApis::OrtGraph_GetSubGraph` calls in the case of the multi-layer nested control-flow op. However, building subgraph context is quick, and most of the subgraphs are not that large.